8284849: Add deoptimization to unified logging

jdksjolen opened the pull request:
This PR adds a new "deoptimization" tag and a new log message on the info level mimicking JFR's deoptimization event. In the future this tag will be used to add trace level information, hopefully replacing TraceDeoptimization. JFR's event also captures the stacktrace, which this does not (happy to add this if considered important).

An example output looks like this: 

```
[1.324s][debug][deoptimization] cid=236 level=4java.util.concurrent.locks.ReentrantLock$Sync.tryRelease(I)Z trap_bci=26 unstable_if reinterpret pc=0x00007f8b30bf8b94 relative_pc=0x0000000000000174
```

Passes tier1.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8284849](https://bugs.openjdk.org/browse/JDK-8284849): Add deoptimization to unified logging


### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [4f488d95](https://git.openjdk.org/jdk/pull/8812/files/4f488d95f7116d0c39c4057408c3bc0714bf2739)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - Committer) ⚠️ Review applies to [864460f9](https://git.openjdk.org/jdk/pull/8812/files/864460f9e483c632bcab954d768abd81201de716)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8812/head:pull/8812` \
`$ git checkout pull/8812`

Update a local copy of the PR: \
`$ git checkout pull/8812` \
`$ git pull https://git.openjdk.org/jdk pull/8812/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8812`

View PR using the GUI difftool: \
`$ git pr show -t 8812`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8812.diff">https://git.openjdk.org/jdk/pull/8812.diff</a>

</details>

JornVernee commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1929,6 +1932,19 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
 
     JFR_ONLY(post_deoptimization_event(nm, trap_method(), trap_bci, trap_bc, reason, action);)
 
+    if(Log(deoptimization)::is_info()) {
```
I don't see this pattern used for checking whether a log tag is enabled used elsewhere. Won't this only print when the log level is exactly `info`?

The pattern I see used elsewhere is:

```
LogTarget(Info, <tag>) lt;
if (lt.is_enabled()) {
    ...
}
```

jdksjolen replied:
This does the same thing. `is_enabled` calls `is_level` internally. I did it this way so that we have a temporary value instead of cluttering the outer scope.

dholmes-ora replied:
Using a `LogMessage` instead of a `LogTarget` is very rarely used in the codebase. I have no idea what the pros and cons of each are, nor can I find any guidance on which to use when.

dholmes-ora replied:
Also style nit: space after if please.

JornVernee commented:
Thanks for looking at this!

I suggest trying to match the style of the `-Xlog:jit+compilation=debug` messages, since deoptimization is like the dual of compilation. And trying to keep the output on a single line,

Rather than printing the compiler name, I think printing the compilation level is more useful, as well as including some info about whether it was an osr compilation maybe. The `jit+compilation` log messages do the latter by printing an extra `%`.

Maybe something like:

```
<compile id> <is osr> <compile level> <method signature> @ <bci> <reason> -> <action>
```

The instruction and line number could also be incorporated somewhere, but the compilation messages don't have them either. They seem fine to leave out I think.

Also, I noticed that `jit+compilation` has the `debug` level, so maybe this log message should be under the `debug` level as well, instead of `info`. I don't know if there is any guidance on that somewhere.

dholmes-ora commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1929,6 +1932,19 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
 
+    if(Log(deoptimization)::is_info()) {
+      LogMessage(deoptimization) lm;
+      auto tm = trap_method();
```
Not sure use if C++11 type inference is a permitted use - I can't find it n the style guide.

jdksjolen replied:
`LogMessage` ensures that messages are kept in order in the final output, but this is not needed if we're logging on one line as Jorn has suggested. I'm also removing the type inference.

kimbarrett replied:
[not a review, just responding to @dholmes-ora question]

Search for "Type Deduction" in the style guide.  There aren't a lot of uses, and the guidance is to only use it when there's significant benefit, not just as a convenience to avoid writing the type.

dholmes-ora commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1929,6 +1932,19 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
+    if(Log(deoptimization)::is_info()) {
+      LogMessage(deoptimization) lm;
+      auto tm = trap_method();
+      lm.info("Deoptimization start {");
```
Do we use this brace style for other log messages? I'm not really seeing the need for a start/end delimiter here.

jdksjolen commented:
> Thanks for looking at this!
> 
> I suggest trying to match the style of the `-Xlog:jit+compilation=debug` messages, since deoptimization is like the dual of compilation. And trying to keep the output on a single line (less output makes it easier to sift through lengthy logs).
> 
> Rather than printing the compiler name, I think printing the compilation level is more useful, as well as including some info about whether it was an osr compilation maybe. The `jit+compilation` log messages do the latter by printing an extra `%`.
> 
> Maybe something like:
> 
> ```
> <compile id> <is osr> <compile level> <method signature> @ <bci> <reason> -> <action>
> ```
> 
> The instruction and line number could also be incorporated somewhere, but the compilation messages don't have them either. They seem fine to leave out I think. (not all class files have useful line numbers, and the bci should be enough to find the bytecode source of the deopt, and from that the instruction).

I'm not sure whether we can know if something was an OCR compilation or not. This is how jit+compilation does it:

```
// CompileTask::print_compilation
void CompileTask::print(outputStream* st, const char* msg, bool short_form, bool cr) {
  bool is_osr_method = osr_bci() != InvocationEntryBci;
  print_impl(st, is_unloaded() ? NULL : method(), compile_id(), comp_level(), is_osr_method, osr_bci(), is_blocking(), msg, short_form, cr, _time_queued, _time_started);
}
```

I don't think we have a `CompileTask` to look at. Other than that, I'm completely up for changing this to a single log line with a similar format to the one you suggested.

> 
> Also, I noticed that `jit+compilation` has the `debug` level, so maybe this log message should be under the `debug` level as well, instead of `info`. I don't know if there is any guidance on that somewhere.

There's some guidance in the JEP, basically `debug` is allowed to be a bit more computationally intensive than `info`. I think it sounds good to move it to `info`.

JornVernee commented:
> I'm not sure whether we can know if something was an OCR compilation or not.

`CompiledMethod` has the function `is_osr_method` which I think can be used for that. The code in question seems to be dealing with a `CompilerMethod*` already (https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/deoptimization.cpp#L1876)

jdksjolen commented:
I changed the log to one line and moved out all repeatedly used values into variables. I limited all of these to a scope to not pollute the function scope unnecessarily. I use `comp_level` directly instead of translating into a string (didn't find a function for it), are levels as ints sufficient?

JornVernee reviewed (commented):
I think this looks good overall. 1 comment inline

jdksjolen commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1927,12 +1930,27 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
+      const char* reason_name = Deoptimization::trap_reason_name(reason);
+      const char* reason_action = Deoptimization::trap_action_name(action);
+      JFR_ONLY(post_deoptimization_event(nm, tm, trap_bci, trap_bc, reason, action);)
+      if (Log(deoptimization)::is_info()) {
```
100% correct, it also checks the wrong log level so double whammy mistake right there.

dholmes-ora commented:
My comments were only "drive-by" observations. You really need the compiler folk to perform the actual review here.

vnkozlov reviewed (commented):
Consider moving this code into separate method.
This part of code is used for processing only uncommon traps when compiled code traps.
A compiled method could be deoptimized for other reasons too. Both paths calls  `fetch_unroll_info_helper()`. May be it is better place to print and log (an other RFE).

jdksjolen commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1927,12 +1930,26 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
+      Method* tm = trap_method();
+      char* name_sig = tm->name_and_sig_as_C_string();
+      const char* reason_name = Deoptimization::trap_reason_name(reason);
+      const char* reason_action = Deoptimization::trap_action_name(action);
```
Thanks, removed :-).

jdksjolen commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1927,12 +1930,26 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
+      JFR_ONLY(post_deoptimization_event(nm, tm, trap_bci, trap_bc, reason, action);)
+      log_debug(deoptimization)("cid: %d osr: %s level: %d %s @ %d %s -> %s",
+                                nm->compile_id(),
+                                (nm->is_osr_method() ? "true" : "false"),
```
This is the current log line (w/o "osr"), it's become quite long.

`[12,240s][debug][deoptimization] cid=135  pc=0x00007fa4dcbef5d4 relative_pc=0x0000000000000154 level=4 java.util.concurrent.locks.ReentrantLock$Sync::tryRelease @ 26 unstable_if reinterpret`

I changed to `=` to make it consistent with the latter example you posted. `osr_bci` is added after the `trap_bci`. I also refactored logging into using `LogStream` in a separate method, but the common logging block is still inline.

vnkozlov commented on src/hotspot/share/runtime/deoptimization.hpp:
```diff
@@ -164,8 +165,11 @@ class Deoptimization : AllStatic {
   private:
-  // Does the actual work for deoptimizing a single frame
-  static void deoptimize_single_frame(JavaThread* thread, frame fr, DeoptReason reason);
+    static void print_ul(CompiledMethod* nm, intptr_t pc, frame& fr, int trap_bci,
+                         const char* reason_name, const char* reason_action);
+
+    // Does the actual work for deoptimizing a single frame
+    static void deoptimize_single_frame(JavaThread* thread, frame fr, DeoptReason reason);
```
Please, keep original indent for this lines to match other line around. Actually `private:` has wrong additional space.

vnkozlov reviewed (commented):
Few more comments.

jdksjolen commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1834,6 +1838,27 @@ static void post_deoptimization_event(CompiledMethod* nm,
+    LogStream ls(lt);
+    ls.print("cid=%d %s pc=" INTPTR_FORMAT " relative_pc=" INTPTR_FORMAT " level=%d",
+             nm->compile_id(), (is_osr ? "%" : ""), pc, fr.pc() - nm->code_begin(),
+             nm->comp_level());
```
>I may be was not clear but do not use % for OSR marking.

My bad, I misunderstood you.

>Please, print pc and relative_pc at the end of output.

Fixed.

jdksjolen commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1834,6 +1838,27 @@ static void post_deoptimization_event(CompiledMethod* nm,
+             nm->comp_level());
+    nm->method()->print_short_name(&ls);
+    ls.print(" @ ");
+    ls.print("%d ", trap_bci);
```
Fixed.

vnkozlov reviewed (approved):
Looks good.
I don't see GHA testing. What testing you did? I will run tier1 in our system to see if it build on all our systems.

vnkozlov commented:
I noticed that method name is missing signature (arguments type) which we have in `TraceDeoptimization` output:
```
UNCOMMON TRAP method=java.util.Properties.getProperty(Ljava/lang/String;)Ljava/lang/String;  bci=10 pc=0x00007fbae0c23a94, relative_pc=0x00000000000000b4, debug_id=0 compiler=c2 compile_id=9 (@0x00007fbae0c23a94) thread=2203929 reason=null_check action=make_not_entrant unloaded_class_index=-1 debug_id=0
```
Yes, `PrintCompilation` does not have it but it would be nice to have it in this output.

vnkozlov commented:
Please, also update the output example in PR's description. It is misleading now.

navyxliu commented:
hi, @jdksjolen , 
Could you update PR description example?  It looks like you have changed it to a one-liner.  just for a record in case we need to revisit this PR.  

Do you consider 'deopt' instead of 'deoptimization'? I feel deoptimization is too long as a tag.  Neither of them is a word anyway. :)

jdksjolen commented:
I reverted to using name_and_sig for output, updated the PR and fixed the nits.

@navyxliu, I did consider moving to `deopt` but opted against it as there are just as long tags in the rest of the set (for example "defaultmethods" and "methodcomparator").

jdksjolen commented:
@vnkozlov, I'm running it on Mach5 Tier1 as I'm writing this comment. Will edit the original PR when it passes.

vnkozlov commented:
> @navyxliu, I did consider moving to `deopt` but opted against it as there are just as long tags in the rest of the set (for example "defaultmethods" and "methodcomparator").

Technically speaking it should be `uncommon_trap_deopt` because it does not cover other case of deoptimization when compiled method deoptimized due to a class change (new related class loading, unloaded, redefined) or re-compilation (from C1 to C2). That path [fetch_unroll_info()](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/deoptimization.cpp#L163) is not covered by this change (may be another RFE).

navyxliu reviewed (approved):
LGTM. I am not a reviewer. we need other reviewers to approve it.

coleenp commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1834,6 +1838,26 @@ static void post_deoptimization_event(CompiledMethod* nm,
 
 #endif // INCLUDE_JFR
 
+void Deoptimization::print_ul(CompiledMethod* nm, Method* tm, intptr_t pc, frame& fr, int trap_bci,
```
From what I can see, this function only called in this cpp file so it only needs to be defined only in this cpp file.  It doesn't seem to use any Deoptimization class members so doesn't need to be a class member.  Then you can revert changes to deoptimize.hpp.

coleenp replied:
And can you call it something like log_deopt(), since the "print" function is generally a function that just prints to an outputStream without including logic about logging.

jdksjolen replied:
Sounds good, fixed.

coleenp commented:
Sorry for the late comments.

jdksjolen commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1834,6 +1838,26 @@ static void post_deoptimization_event(CompiledMethod* nm,
+    LogStream ls(lt);
+    bool is_osr = nm->is_osr_method();
+    ls.print("cid=%d%s level=%d",
+             nm->compile_id(), (is_osr ? " osr" : ""), nm->comp_level());
```
Sounds good, fixed.

coleenp commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -1834,6 +1838,26 @@ static void post_deoptimization_event(CompiledMethod* nm,
 
 #endif // INCLUDE_JFR
 
+void log_deopt(CompiledMethod* nm, Method* tm, intptr_t pc, frame& fr, int trap_bci,
```
Add 'static' keyword so it has internal linkage.

jdksjolen replied:
Fixed.

coleenp commented on src/hotspot/share/runtime/deoptimization.hpp:
```diff
@@ -44,7 +44,8 @@ class Deoptimization : AllStatic {
   friend class VMStructs;
   friend class EscapeBarrier;
 
- public:
```
You can revert changes to this file since this wasn't part of your change.  Somebody can fix this some other time when they're changing this file.

coleenp reviewed (changes requested):
Just a couple of small tweaks. Thanks.

coleenp reviewed (approved):
Perfect!  Thank you.

vnkozlov reviewed (approved):
Latest version is good.

jdksjolen commented:
/integrate

I'll need a sponsor on this too!

JornVernee commented:
/sponsor
